app/src/androidTest/java/sdp/moneyrun/DatabaseProxyTest.java

```diff
@@ -42,4 +53,121 @@ public void onComplete(@NonNull Task<DataSnapshot> task) {
         }
     }
 
+    @Test(expected = IllegalArgumentException.class)
+    public void addPlayerListenerThrowsExceptionOnNullPlayer() {
+        DatabaseProxy db = new DatabaseProxy();
+        ValueEventListener listener = new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot snapshot) {
+                return;
+            }
+
+            @Override
+            public void onCancelled(@NonNull DatabaseError error) {
+
+            }
+        };
+        db.addPlayerListener(null, listener);
+    }
+
+    @Test(expected = IllegalArgumentException.class)
+    public void addPlayerListenerThrowsExceptionOnNullListener() {
+        DatabaseProxy db = new DatabaseProxy();
+        Player player = new Player(1);
+        db.addPlayerListener(player, null);
+    }
+
+    @Test
+    public void addPlayerListenerCorrectlyUpdatesData(){
+        Player player = new Player(564123);
+        player.setAddress("FooBarr");
+        player.setName("Johann");
+        final DatabaseProxy db = new DatabaseProxy();
+        db.putPlayer(player);
+        String newName = "Simon";
+        try {
+            Thread.sleep(1000);
+        } catch (InterruptedException e) {
+            e.printStackTrace();
+        }
+        db.addPlayerListener(player, new ValueEventListener() {
+            @Override
+            public void onDataChange(@NonNull DataSnapshot snapshot) {
```

Comment:
Does this work in real time? That is when an activity is launched?

Reply:
What do you mean when an activity is launched ? This work in real time yes and without Thread.sleep

Reply:
I mean does it work when called from an activity? Like for example would we be able to do this in the MainActivity in any of it's cycles (onCreate, onResume, onDestroy...)?

Reply:
I sure hope so. I had tested it with a dummy activity last week and it worked fine